src/v/iceberg/merge_append_action.cc

```diff
@@ -0,0 +1,552 @@
+    const auto& pspec = *pspec_it;
+    for (const auto& f : new_data_files_) {
+        if (f.partition.val == nullptr) {
+            vlog(
+              log.error,
+              "Metadata for data file {} is missing partition key",
+              f.file_path);
+            co_return action::errc::unexpected_state;
+        }
+        auto f_num_fields = f.partition.val->fields.size();
+        if (f_num_fields != pspec.fields.size()) {
+            vlog(
+              log.error,
+              "Partition key for data file {} has {} fields, expected {}",
+              f.file_path,
+              f_num_fields,
+              pspec.fields.size());
+            co_return action::errc::unexpected_state;
+        }
+    }
+
+    // Get the manifest list for the current snapshot, if any.
+    manifest_list mlist;
+    std::optional<snapshot_id> old_snap_id;
+    if (table_.snapshots.has_value() && !table_.snapshots->empty()) {
+        if (!table_.current_snapshot_id.has_value()) {
+            // We have snapshots, but it's unclear which one to base our update
+            // off of.
+            vlog(
+              log.error,
+              "Table's current snapshot id is not set but there are {} "
+              "snapshots",
+              table_.snapshots->size());
+            co_return action::errc::unexpected_state;
+        }
+        // Look for the current snapshot.
+        const auto table_cur_snap_id = *table_.current_snapshot_id;
+        const auto& snaps = *table_.snapshots;
+        auto snap_it = std::find_if(
+          snaps.begin(), snaps.end(), [table_cur_snap_id](const snapshot& s) {
+              return table_cur_snap_id == s.id;
+          });
+        if (snap_it == snaps.end()) {
+            // We have snapshots, but the one we thought we needed to base our
+            // update off of is missing.
+            vlog(
+              log.error,
+              "Table's current snapshot id {} is missing",
+              table_cur_snap_id);
+            co_return action::errc::unexpected_state;
+        }
+        auto mlist_res = co_await io_.download_manifest_list(
+          manifest_list_path{snap_it->manifest_list_path});
+        if (mlist_res.has_error()) {
+            co_return to_action_errc(mlist_res.error());
+        }
+        mlist = std::move(mlist_res).value();
+        old_snap_id = table_cur_snap_id;
+    } else if (table_.current_snapshot_id.has_value()) {
+        vlog(
+          log.error,
+          "Table's current snapshot id is set to {} but there are no "
+          "snapshots",
+          table_.current_snapshot_id.value());
+        co_return action::errc::unexpected_state;
+    }
+    const auto new_seq_num = sequence_number{table_.last_sequence_number() + 1};
+    const auto new_snap_id = generate_unused_snap_id(table_);
+
+    // TODO: support more than one partition spec by grouping the merged
+    // manifests by spec id.
+    auto pk_type = partition_key_type::create(pspec, schema);
```

Comment:
i don't know what all assumptions are made about concurrent access, but `schema` for example, is an iterator reference to action state held across `co_await` calls

Reply:
The assumption for actions is that there are no concurrent method calls in a given transaction. This means table_ is expected to be constant throughout this method